arangod/VocBase/Methods/Indexes.cpp

```diff
@@ -545,143 +622,77 @@ arangodb::Result Indexes::createIndex(LogicalCollection* coll,
-    collectionName = std::string(handle.data(), split);
-    iid = IndexId{StringUtils::uint64(handle.data() + split + 1,
-                                      handle.size() - split - 1)};
-    return true;
-  }
-
-  if (!handle.empty() &&
-      !Index::validateId(std::string_view(handle.data(), handle.size()))) {
-    return false;
-  }
-  iid = IndexId{StringUtils::uint64(handle.data(), handle.size())};
-  return true;
-}
-
-////////////////////////////////////////////////////////////////////////////////
-/// @brief checks if argument is an index name
-////////////////////////////////////////////////////////////////////////////////
-
-static bool ExtractIndexName(VPackSlice const& arg, bool extendedNames,
-                             std::string& collectionName, std::string& name) {
-  TRI_ASSERT(collectionName.empty());
-  TRI_ASSERT(name.empty());
-
-  if (!arg.isString()) {
-    return false;
-  }
-
-  std::string_view handle = arg.stringView();
-  if (arangodb::Index::validateHandleName(extendedNames, handle)) {
-    std::size_t split = handle.find('/');
-    TRI_ASSERT(split != std::string::npos);
-    collectionName = std::string(handle.data(), split);
-    name = std::string(handle.data() + split + 1, handle.size() - split - 1);
-    return true;
-  }
-
-  if (IndexNameValidator::isAllowedName(extendedNames, handle)) {
-    name = std::string(handle.data(), handle.size());
-    return true;
-  }
-  return false;
-}
-
 ////////////////////////////////////////////////////////////////////////////////
 /// @brief looks up an index identifier
 ////////////////////////////////////////////////////////////////////////////////
 
-Result Indexes::extractHandle(arangodb::LogicalCollection const* collection,
+Result Indexes::extractHandle(arangodb::LogicalCollection const& collection,
                               arangodb::CollectionNameResolver const* resolver,
                               VPackSlice const& val, IndexId& iid,
                               std::string& name) {
   // reset the collection identifier
   std::string collectionName;
 
-  // assume we are already loaded
-  TRI_ASSERT(collection != nullptr);
-
-  bool extendedNames = collection->vocbase()
+  bool extendedNames = collection.vocbase()
                            .server()
                            .getFeature<DatabaseFeature>()
-                           .extendedNamesForCollections();
+                           .extendedNames();
 
   // extract the index identifier from a string
   if (val.isString() || val.isNumber()) {
-    if (!ExtractIndexHandle(val, extendedNames, collectionName, iid) &&
-        !ExtractIndexName(val, extendedNames, collectionName, name)) {
-      return Result(TRI_ERROR_ARANGO_INDEX_HANDLE_BAD);
+    Result res = ::extractIndexHandle(val, extendedNames, collectionName, iid);
+    if (!res.ok()) {
+      res = ::extractIndexName(val, extendedNames, collectionName, name);
+    }
+    if (!res.ok()) {
```

Comment:
I don't think so.
The code is intended to work as follows:
```
res = findIndexHandleInData();
if (!res.ok()) { // no index handle found
  res = findIndexNameInData();
}
if (!res.ok()) {
  return ...
}
```
So in case we don't find a handle, `res.ok()` is false, but we still want to find the index name in the data.
Only if both approaches fail, we return the error.
I have added code comments.